staging/src/k8s.io/client-go/tools/cache/reflector_test.go

```diff
@@ -358,6 +360,80 @@ func TestReflectorListAndWatchWithErrors(t *testing.T) {
 	}
 }
 
+func TestReflectorListAndWatchInitConnBackoff(t *testing.T) {
+	maxBackoff := 50 * time.Millisecond
+	table := []struct {
+		numConnFails  int
+		expLowerBound time.Duration
+		expUpperBound time.Duration
+	}{
+		{5, 32 * time.Millisecond, 64 * time.Millisecond}, // case where maxBackoff is not hit, time should grow exponentially
+		{40, 35 * 2 * maxBackoff, 40 * 2 * maxBackoff},    // case where maxBoff is hit, backoff time should flatten
+
+	}
+	for _, test := range table {
+		t.Run(fmt.Sprintf("%d connection failures takes at least %d ms", test.numConnFails, 1<<test.numConnFails),
+			func(t *testing.T) {
+				stopCh := make(chan struct{})
+				connFails := test.numConnFails
+				fakeClock := clock.NewFakeClock(time.Unix(0, 0))
+				bm := wait.NewExponentialBackoffManager(time.Millisecond, maxBackoff, 100*time.Millisecond, 2.0, 1.0, fakeClock)
+				var iter int
+				go func() {
+					for {
+						if fakeClock.HasWaiters() {
+							step := (1 << (iter + 1)) * time.Millisecond
```

Comment:
do you ever increment iter?

Reply:
yup, we decrement `connFails` on every iteration of `WatchFunc` and recalculate `iter` as the difference between the number of times the connection should fail (`test.numConnFailes`) and the number of times it has already failed (`connFails`)

https://github.com/kubernetes/kubernetes/pull/94235/files#diff-4713f63cf917b7881fa12f47f3d9ec75R395-R397

Reply:
hm, I think that is a data race?

Reply:
Added a mutex

Reply:
Another option might be to send a new value down a channel when it is recomputed, or stop and re-launch this goroutine.

Alternatively, can you just increment iter every time this logic triggers?

Also, you want some way for the goroutine to exit, otherwise this test will just leak a few goroutines every time it runs.

Reply:
Ah yea, I think separating it from connFails and just bumping it every time we step the clock makes more sense.

I've added a return to the go routine via a select statement on a done channel. Let me know if it's not idiomatic.